### PR TITLE
[CS2] Compile computed properties to ES2015 equivalent

### PR DESCRIPTION
**Pull request by connec**
This is a fairly small change that simplifies the code generation for computed properties as they're now generated in the object initializer like regular properties.

I ended up doing this whilst working on fully switching to ES2015 splats. Targets the `2` branch since the output won't compile without ES2015 support.

- [x] Tests pass (excepting #4336)
- [x] Compiles the CoffeeScript code base without modifications
- [ ] Compiles a codebase making significant use of computed keys - any suggestions?

### CODE REVIEW

**Comment by connec** [comment]
Also, the first commit is actually just compiling some files in `2` that weren't committed for some reason.

File: src/nodes.coffee [hunk 1]
```diff
@@ -981,54 +981,37 @@ exports.Obj = class Obj extends Base
...
         if prop.variable instanceof Value and prop.variable.hasProperties()
           prop.variable.error 'invalid object key'
       if prop instanceof Value and prop.this
         prop = new Assign prop.properties[0].name, prop, 'object'
-      if prop not instanceof Comment
-        if i < dynamicIndex
-          if prop not instanceof Assign
-            prop = new Assign prop, prop, 'object'
+      if prop not instanceof Comment and prop not instanceof Assign
+        if prop.isComplex()
```
Comment (connec) [inline src/nodes.coffee]: Previously, dynamic keys were being detected with...
```coffee
(prop.variable or prop).base instanceof Parens
```
... but this seems simpler. I don't know all the ins and outs though, so I might be missing something.

**Comment by jashkenas** [comment]
Nice stuff!

**Comment by GeoffreyBooth** [comment]
@connec this looks good, do you mind just giving an example of some CoffeeScript with computed properties, and what it was compiling to before and after this PR?

**Comment by lydell** [comment]
@GeoffreyBooth Here's an example: https://github.com/jashkenas/coffeescript/pull/3840/commits/76c076db555c9ac7c325c3b285cd74644a9bf0d2

**Comment by connec** [comment]
Sure:

Current (1.11.1):
```
★ coffee -bcs
{
  "#{computed}": 1
  "#{f()}"
  normal: 2
  "#{a.bit.more.complex = computation()}": 3
}
// Generated by CoffeeScript 1.11.1
var obj, ref;

(
  obj = {},
  obj["" + computed] = 1,
  obj[ref = "" + (f())] = ref,
  obj.normal = 2,
  obj["" + (a.bit.more.complex = computation())] = 3,
  obj
);
```

In this PR:

```
★ bin/coffee -bcs
{
  "#{computed}": 1
  "#{f()}"
  normal: 2
  "#{a.bit.more.complex = computation()}": 3
}
// Generated by CoffeeScript 1.11.0
var ref;

({
  ["" + computed]: 1,
  [ref = "" + (f())]: ref,
  normal: 2,
  ["" + (a.bit.more.complex = computation())]: 3
});
```

**Comment by GeoffreyBooth** [comment]
This looks good to me. I updated your branch with the latest `2`, and the tests pass. Anyone opposed to merging this in?

**Comment by lydell** [comment]
Go for it!

**Comment by danielbayley** [comment]
Following on from #3597, wouldn't it now be preferable to simply allow the existing ES6 `[key]: value` syntax and pass it straight through for this one?

**Comment by connec** [comment]
@danielbayley the purpose wasn't to add or change any syntax, interpolations were already allowed in object keys, this just simplifies the output by compiling to ES6. Might be worth having a separate discussion about `[key]` syntax.

Personally I prefer CS' approach to this. If "hello" is a valid key, then so too should be "#{hello}". ES' handling of template strings is needlessly restrictive imo.

**Comment by danielbayley** [comment]
> interpolations were already allowed in object keys

Is this what you were referring to?
~~~ coffee
key = 'key'

obj = {}
obj[key] = 'value'
~~~
 Obviously this is nothing new, but just to clarify…
~~~ coffee
"#{key}": 'value' # works… cool! Leave it be.

[key]: 'value' # is valid ES6 but I think should also work in CS.
~~~
Maybe you're right about it being a separate discussion; is it worth opening another issue?

**Comment by connec** [comment]
>> interpolations were already allowed in object keys
>
> @connec Is this what you were referring to?

Sorry, I just meant that this syntax was valid before the PR:

```coffee
key = 'key'
obj =
  "#{key}": 'value'
```

It used to compile to (roughly):

```js
var key, obj

key = 'key';

obj = (
  ref = {},
  ref['' + key] = 'value',
  ref
);
```

And now compiles to:

```js
var key, obj

key = 'key';

obj = {
  ['' + key]: 'value'
};
```

> is it worth opening another issue?

I'm not sure, @GeoffreyBooth may have an opinion on this.

**Comment by GeoffreyBooth** [comment]
Personally I don’t see why `[key]: 'value'` shouldn’t be valid CoffeeScript. @lydell? I think this corresponds with our discussing allowing `...` on either side of a variable name, e.g. `...args` or `args...`, to ease the transition for people who write lots of ES.

Perhaps open a new issue requesting the new syntax. Please prefix it with `[CS2]`, as I don’t think we’re revisiting this anymore in 1.x.

**Comment by lydell** [comment]
I guess we more or less have to support `[key]: 'value'` to allow stuff like `[Symbol.iterator]: foo`.

**Comment by connec** [comment]
> `[Symbol.iterator]: foo`

Good point 😢

I still think it's consistent to allow interpolations in keys though.

**Comment by lydell** [comment]
> I still think it's consistent to allow interpolations in keys though.

Me too.

**Pull request merged**